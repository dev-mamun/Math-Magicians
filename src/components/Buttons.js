/** ****************************************
 * Project: math-magicians
 * File: Buttons.js
 * Created: 3/29/23
 * Author: Abdullah Al Mamun <mamun1214@gmail.com>
 ****************************************** */
import React, { useState } from 'react';
import calculate from '../logic/calculate';
import InputFied from './Input';

function Buttons() {
  const [display, setDisplay] = useState('0');
  const [calObj, setCalObj] = useState({
    total: null,
    next: null,
    operation: null,
  });
  const btnHandler = (buttonName) => {
    const $data = calculate(calObj, buttonName.toString());
    setCalObj($data);

    // Update the display value
    const $total = $data.total || '';
    const $next = $data.next || '';
    const $operator = $data.operation || '';
    const $display = $total + $operator + $next;
    setDisplay($display);
  };
  return (
    <>
      <InputFied display={display} />
      <div className="buttons">
        <div className="btn-row">
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('AC')}
            >
              AC
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('+/-')}
            >
              +/-
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('%')}
            >
              %
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('รท')}
            >
              รท
            </button>
          </div>
        </div>
        <div className="btn-row">
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('7')}
            >
              7
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('8')}
            >
              8
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('9')}
            >
              9
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('x')}
            >
              *
            </button>
          </div>
        </div>
        <div className="btn-row">
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(4)}
            >
              4
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(5)}
            >
              5
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(6)}
            >
              6
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('-')}
            >
              -
            </button>
          </div>
        </div>
        <div className="btn-row">
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(1)}
            >
              1
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(2)}
            >
              2
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(3)}
            >
              3
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('+')}
            >
              +
            </button>
          </div>
        </div>
        <div className="btn-row">
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler(0)}
            >
              0
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('.')}
            >
              .
            </button>
          </div>
          <div className="btn-col">
            <button
              type="button"
              className="cal-btn"
              onClick={() => btnHandler('=')}
            >
              =
            </button>
          </div>
        </div>
      </div>
    </>
  );
}

export default Buttons;
