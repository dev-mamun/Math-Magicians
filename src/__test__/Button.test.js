/** ****************************************
 * Project: math-magicians
 * File: Button.test.js
 * Created: 4/8/23
 * Author: Abdullah Al Mamun <mamun1214@gmail.com>
 ****************************************** */
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Buttons from '../components/Buttons';

describe('Buttons are rendered correctly', () => {
  test('Buttons snapshots match', () => {
    const tree = renderer.create(<Buttons />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('ensure all buttons are rendered', () => {
    render(<Buttons />);
    const buttons = screen.getAllByRole('button');
    expect(buttons.length).toBe(19);
  });
});