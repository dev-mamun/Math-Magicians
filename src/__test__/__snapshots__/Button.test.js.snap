// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons are rendered correctly Buttons snapshots match 1`] = `
Array [
  <div
    className="dispaly"
  >
    <input
      className="input"
      onChange={[Function]}
      placeholder="0"
      type="text"
      value="0"
    />
  </div>,
  <div
    className="buttons"
  >
    <div
      className="btn-row"
    >
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          %
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          รท
        </button>
      </div>
    </div>
    <div
      className="btn-row"
    >
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          7
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          8
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          9
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          *
        </button>
      </div>
    </div>
    <div
      className="btn-row"
    >
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          6
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
    <div
      className="btn-row"
    >
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
    </div>
    <div
      className="btn-row"
    >
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          .
        </button>
      </div>
      <div
        className="btn-col"
      >
        <button
          className="cal-btn"
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>,
]
`;
